scheduler_simulator.py

```diff
@@ -395,12 +395,17 @@ def run(self, until: float = float('inf')):
 
 
 class RoundRobin(Scheduler):
-    """ Chooses an available path in a round-robin manner between two paths. """
+    """ Chooses an available path in a round-robin manner between multiple paths. """
     last_path: Optional[Path] = None
 
     def schedule(self, packet_len: int) -> Optional[Path]:
-        for p in self.paths:
-            if p != self.last_path and not p.blocked(packet_len):
+        try:
+            next = self.paths.index(self.last_path) + 1
+        except ValueError:
+            next = 0
+        sorted_paths = self.paths[next:] + self.paths[:next]
```

Comment:
I like the way `sorted_paths` is expressed, but I would reformulate `next` as something like:
`next_idx = (self.paths.index(self.last_path) + 1) % len(self.paths) if self.last_path else 0`
Storing the last index instead of the last path makes it even easier if you initialize it to -1 instead of None:
```
last_path: int = -1
...
next_idx = (self.last_path_idx + 1) % len(self.paths)
sorted_paths = self.paths[next_idx:] + self.paths[:next_idx]
```
Then when updating `last_path_idx` we can use the index function and the selected path.

Reply:
Thank you for the review!

> but I would reformulate next as something like: next_idx (...)

no problem to rename `next` to `next_idx`. I even prefer and I simply tried to keep variables with a short name like in the other examples.

> next_idx = (self.paths.index(self.last_path) + 1) % len(self.paths) if self.last_path else 0

I don't think the modulo is needed (`% len(self.paths)`):

```python
>>> l = [1, 2, 3, 4]
>>> l[4:]
[]
>>> l[:4]
[1, 2, 3, 4]
```

Initially I wrote:

    next = self.paths.index(self.last_path) + 1 if self.last_path in self.paths else 0

but the line was long. I can change if you prefer, no problem for me

> Storing the last index instead of the last path makes it even easier if you initialize it to -1 instead of None

I agree with you that it is easier here. It is just that if the list of paths is modified, `last_path_idx` might not be valid any more. I don't know how the simulator will be extended in the future but I can change. If we will never do that, I am perfectly fine if we keep the index.

Reply:
> I don't think the modulo is needed (% len(self.paths)):

Great, sorry for that.
> It is just that if the list of paths is modified, last_path_idx might not be valid any more. I don't know how the simulator will be extended in the future but I can change

Correct. Me neither, let's keep it your way.

> no problem to rename next to next_idx

That's a typo on my side, not a real comment. I don't mind any of the two.

Ok then, maybe we could replace the try except with an if else based on the condition `self.last_path in self.paths`. That's a bit easier to read, isn't it ?

Reply:
> Great, sorry for that.

No problem!

> Correct. Me neither, let's keep it your way.

OK, up to you. We can change that in a next version if required.

> That's a typo on my side, not a real comment. I don't mind any of the two.

OK

> Ok then, maybe we could replace the try except with an if else based on the condition self.last_path in self.paths. That's a bit easier to read, isn't it ?

I would find it easier with a real "if/else"  :)
but let's move it on one line!